[stable/prometheus] RBAC, namespaces, and node-exporter default update

alejandroEsc opened the pull request:
- rbac is off by default, but can be turned on
- added clusterroles, bindings and service accounts
- added namespaces where needed.
- updated default node-exporter because previous version had a fairly serious bug disallowing it to start up properly.

alejandroEsc committed:
first commit to add rbac to prometheus.
added clusterroles, bindings and service accounts

added configuration changes to the readme doc.

updated cluster roles.

additional changes made to simplify rbac reasoning

updated the readme file.

addressed PR items.

alejandroEsc commented:
@mgoodness could you please review or assign?

alejandroEsc commented:
@mgoodness could I get an eta on this review? if you are too busy any suggestions to possible maintainers that may be able to take over?

cortesi commented:
@mgoodness I'd like to add my voice to @alejandroEsc here. The Prometheus chart is critical, and it is often broken for months on end with all PRs self-assigned to you and no activity.

mgoodness commented on stable/prometheus/templates/alertmanager-deployment.yaml:
```diff
@@ -22,6 +22,9 @@ spec:
     spec:
+      {{- if .Values.rbac.install }}
+      serviceAccountName: alert-manager-account
+      {{- end }}
```
Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.alertmanager.fullname" . }}{{ else }}"{{ .Values.alertmanager.serviceAccountName }}"{{ end }}`

mgoodness commented on stable/prometheus/templates/kube-state-metrics-deployment.yaml:
```diff
@@ -22,6 +22,9 @@ spec:
     spec:
+      {{- if .Values.rbac.install }}
+      serviceAccountName: kube-state-metrics-account 
+      {{- end }}
```
Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.kubeStateMetrics.fullname" . }}{{ else }}"{{ .Values.kubeStateMetrics.serviceAccountName }}"{{ end }}`

mgoodness commented on stable/prometheus/templates/node-exporter-daemonset.yaml:
```diff
@@ -21,6 +21,9 @@ spec:
     spec:
+      {{- if .Values.rbac.install }}
+      serviceAccountName: node-exporter-account
+      {{- end }}
```
Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.nodeExporter.fullname" . }}{{ else }}"{{ .Values.nodeExporter.serviceAccountName }}"{{ end }}`

mgoodness commented on stable/prometheus/templates/server-deployment.yaml:
```diff
@@ -21,6 +21,9 @@ spec:
     spec:
+      {{- if .Values.rbac.install }}
+      serviceAccountName: prometheus-server-account
+      {{- end }}
```
Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.fullname" . }}{{ else }}"{{ .Values.server.serviceAccountName }}"{{ end }}`

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+---
```
Delete all YAML separators, and break the file up (one per CRB). Grouping resources like this can cause problems when upgrading.

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+---
+{{- if .Values.rbac.install }}
```
Use `{{- if .Values.createRbac -}}`

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+{{- if .Values.rbac.install }}
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/{{ .Values.rbac.apiVersion }}
```
We only support the last two versions of Kubernetes (currently 1.6 & 1.7), so it's safe to statically set the API version to `v1beta1`.

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/{{ .Values.rbac.apiVersion }}
+metadata:
+  name: prometheus-server-binding
```
Use `name: {{ template "prometheus.fullname" . }}`

Also add standard labels (app, chart, release, component, heritage).

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+  namespace: {{ .Release.Namespace }}
+subjects:
+- kind: ServiceAccount
+  name: prometheus-server-account
```
Use `name: {{ template "prometheus.fullname" . }}`

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+  namespace: {{ .Release.Namespace }}
+roleRef:
+  kind: ClusterRole
+  name: prometheus-server
```
Use `name: {{ template "prometheus.fullname" . }}`

mgoodness commented on stable/prometheus/templates/rbac-clusterrolebindings.yaml:
```diff
@@ -0,0 +1,62 @@
+roleRef:
+  kind: ClusterRole
+  name: prometheus-server
+  #name: cluster-admin
```
Delete

mgoodness reviewed (changes requested):
The changes I'm requesting generally follow the RBAC pattern we established in #1504. I only commented on the first part of `rbac-clusterrolebindings`, but the changes should be applied throughout that file, `rbac-clusterroles`, and `rbac-serviceaccounts`.

mgoodness commented:
Fixes #1075

alejandroEsc commented:
@mgoodness thanks for the review, will start ASAP.

alejandroEsc commented:
@mgoodness I believe I addressed all your points. The only thing I did not address was the apiVersion rbac option. You suggest I leave it as v1beta1 and remove this value. But I think until we are at v1 it still a good idea.

naveensrinivasan commented:
@mgoodness Any update on this?

chainlink commented:
Also waiting on this one to test prometheus on GCE

maartenvanvliet commented:
I have enabled Legacy authorization on GCE to test prometheus. Works great. Waiting for this to be accepted though, and disable Legacy authorization again.

chainlink commented:
Good to know @maartenvanvliet !

mgoodness commented:
Reviewing right now. Sit tight.

mgoodness committed:
Conform to standard RBAC pattern

mgoodness committed:
Bump chart version

mgoodness committed:
Delete extra metadata

mgoodness commented:
serviceAccountNames now defaults to `default` if `rbac.create=false`, as is our pattern.

I also set the RBAC apiVersion to v1beta1. When it hits v1, we'll handle compatibility with a conditional based on the `.Capabilities` object.

Please test one more time to make sure I didn't bork anything.

mgoodness commented:
/lgtm

mgoodness committed:
Delete README whitespace

jurgenweber commented:
eta on merge? This would be really handy.

dhilipkumars commented:
/retest

seh commented on stable/prometheus/templates/alertmanager-clusterrolebinding.yaml:
```diff
@@ -0,0 +1,20 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cluster-admin
```
- [ ] Given that Alertmanager only _reads_ objects, giving it the "cluster-admin" _ClusterRole_ is rather generous.  
  Better would be to define a _ClusterRole_ that allows it to read only what it needs to read.

seh replied:
This still strikes me as dangerous. If an administrator installs this chart into a namespace in which other non-administrators can run pods, those other users can run a pod using this service account, and escalate their privileges up to "cluster-admin." I think you should at least document this hazard if you're not willing to come up with a safer, narrower _ClusterRole_ for Alertmanager here.

mgoodness replied:
Any refinement of the the RBAC permissions can & should be submitted through PRs.

seh commented on stable/prometheus/templates/node-exporter-clusterrolebinding.yaml:
```diff
@@ -0,0 +1,20 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cluster-admin
```
- [ ] Given that Node Exporter only reads things, giving it the "cluster-admin" _ClusterRole_ is rather generous.  
  Better would be to define a _ClusterRole_ that allows it to read only what it needs to read. What does it need to read from the API server?

seh replied:
[Likewise](https://github.com/kubernetes/charts/pull/1443/files#r133989161), but here with Node Exporter.

seh commented on stable/prometheus/values.yaml:
```diff
@@ -1,8 +1,14 @@
   enabled: true
 
+  # Defines the serviceAccountName to associate with a role when rbac is enabled.
+  serviceAccountName: default
```
- [x] You risk granting the "cluster-admin" _ClusterRole_ to the "default" service account here.

mgoodness replied:
Pods only run as `.Values.serviceAccountName` if `rbac.create=false`. In that case, none of the RBAC resources are created and it's up to the user to configure permissions externally.

seh replied:
That comment above is confusing, then. It says, "when [RBAC] is enabled," which I took to mean "when this chart plays along with RBAC."

mgoodness replied:
Fair enough. I'll be explicit about `rbac.create=false`.

mgoodness commented:
@jurgenweber I'd been [waiting](https://github.com/kubernetes/charts/pull/1443#issuecomment-322320396) for someone to deploy & test this before I merged. Also, like @seh, I'm somewhat uncomfortable running alertmanager & node exporter as `cluster-admin`. However, if the consensus is to improve RBAC in another PR, I'll get this merged today.

mgoodness committed:
Clarify RBAC requirements

chainlink commented:
Tested this on gce with 1.6 and it works great out of the box!

jurgenweber commented:
@mgoodness ok, I started giving it a go and then go distracted but as far as I could see all of the resources were green. Not a great test I know, but it is a fresh install for me.

jurgenweber commented:
thanks for the merge, for the record I upgraded to kube 1.7.4 and it works great. :) Thanks again.
